src/filecache.c

```diff
@@ -386,7 +386,7 @@ static void get_fresh_fd(filecache_t *cache,
     // If we're in saint mode, don't go to the server
     if (pdata != NULL &&
             ((flags & O_TRUNC) || use_local_copy ||
-            (pdata->last_server_update == 0) || (time(NULL) - pdata->last_server_update) <= STAT_CACHE_NEGATIVE_TTL)) {
+            (pdata->last_server_update == 0) || pdata->last_server_update > FILECACHE_INVALIDATED)) {
```

Comment:
This seems to make the implicit assumption that there's been a recent PROPFIND invalidation pass on the directory prior to reaching here. While seems reasonable given expected kernel behavior, I'm not sure how I feel about relying on that ordering for correctness. I think I need to give this some thought.

Reply:
I feel we have options to de-risk the change. The plan was always to use valhalla log data to find customers with a high number of 304s and less-than-elite plan/tier and use them for a canary release that would be closely monitored. We could additionally add a softer age out, like 60 seconds or so. That way any unexpected behavior would correct itself fairly quickly, while also letting the invalidation logic have a potentially significant effect on the number of 304s. Files that are actually being updated many times a minute would have problems, but I imagine sites with that behavior aren't performing very well anyway. As we gain confidence we could grow the age out interval.

We could also add another type of cache entry for "when did we last update the cache", which would let us log warnings and bump statsd counters when for some reason we did not update our caches recently before entering `get_fresh_fd`. I would think that, even with races to update this "meta-cache" entry, our worst case result of keeping the oldest update time would give us an update time that's inaccurate by a few seconds, 2 or 3 at most. We could introduce telemetry for this before actually introducing the invalidation logic.

Reply:
How would you feel about checking the directory's last freshening and using the later of the directory and file freshness timestamps?

Reply:
As clarified in slack, I feel great about this. Also, I learned that the stat cache actually has a special entry for this, which why it's even possible to set the query parameter for progressive propfinds. I've rebased on the change that allows me to build docker images and debug them, mostly so I could test this change (cursory testing show that it works).

I'm not explicitly trying for the latter of two timestamps, I'm just setting a grace period for the validity of a directory update that would have invalidated some file cache records, and falling back on the logic that was already there. I'm introducing separate defines for separate concepts in an effort to remove a trap where someone could intentionally change "the time after which a negative entry in the stat cache is no longer valid" and unwittingly change a bunch of other things. All these things are `2` (seconds) ... for now.

Reply:
... and the latest commit is my latest stab at doing this.